packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx

```diff
@@ -55,17 +57,35 @@ export default function useTdpClientCanvas(props: Props) {
     canvas.height = height;
   };
 
-  // Default TdpClientEvent.IMAGE_FRAGMENT handler (buffered)
+  // Default TdpClientEvent.TDP_PNG_FRAME handler (buffered)
   const onPngFrame = (ctx: CanvasRenderingContext2D, pngFrame: PngFrame) => {
-    // The first image fragment we see signals a successful rdp connection on the backend.
-    if (firstImageFragmentRef.current) {
+    // The first image fragment we see signals a successful tdp connection.
+    if (!initialTdpConnectionSucceeded.current) {
       syncCanvasSizeToDisplaySize(ctx.canvas);
       setTdpConnection({ status: 'success' });
-      firstImageFragmentRef.current = false;
+      initialTdpConnectionSucceeded.current = true;
+      // sendLocalClipboardToRemote must be called only after
+      // initialTdpConnectionSucceeded.current === true or else it won't do anything.
+      if (shareClipboard) sendLocalClipboardToRemote(tdpClient);
     }
     ctx.drawImage(pngFrame.data, pngFrame.left, pngFrame.top);
   };
 
+  // Default TdpClientEvent.TDP_CLIPBOARD_DATA handler.
+  const onClipboardData = (clipboardData: ClipboardData) => {
+    if (
+      document.hasFocus() &&
+      clipboardData.data !== latestClipboardText.current
+    ) {
+      navigator.clipboard.writeText(clipboardData.data).then(() => {
+        // Set latestClipboardText.current to whatever we got, so that
+        // next time onMouseEnter fires we don't try to send it back to
+        // the remote machine.
+        latestClipboardText.current = clipboardData.data;
+      });
+    }
+  };
```

Comment:
I think our current system would prevent that from being an issue, but it doesn't hurt to add an extra check in case I'm not thinking of an edge case or the upstream system changes.